hxmath/math/DualQuaternion.hx

```diff
@@ -0,0 +1,622 @@
+    }
+    
+    /**
+     * Copy the contents of this structure to another.
+     * Faster than copyToShape for static platforms (C++, etc) but requires the target to have the exact same inner type.
+     * 
+     * @param target    The target structure.
+     */
+    public inline function copyTo(target:DualQuaternion):Void
+    {
+        target.real.copyTo( this.real );
+        target.dual.copyTo( this.dual );
+    }
+    
+    
+    /**
+     * Get an element by position.
+     * 
+     * @param i         The element index.
+     * @return          The element.
+     */
+    @:arrayAccess
+    public inline function getArrayElement(i:Int):Float
+    {
+        var self:DualQuaternion = this;
+        switch (i)
+        {
+            case 0:
+                return self.real.s;
+            case 1:
+                return self.real.x;
+            case 2:
+                return self.real.y;
+            case 3:
+                return self.real.z;
+            case 4:
+                return self.dual.s;
+            case 5:
+                return self.dual.x;
+            case 6:
+                return self.dual.y;
+            case 7:
+                return self.dual.z;
+            default:
+                throw "Invalid element";
+        }
+    }
+    
+    /**
+     * Set an element by position.
+     * 
+     * @param i         The element index.
+     * @param value     The new value.
+     * @return          The updated element.
+     */
+    @:arrayAccess
+    public inline function setArrayElement(i:Int, value:Float):Float
+    {
+        var self:DualQuaternion = this;
+        switch (i)
+        {
+            case 0:
+                return self.real.s = value;
+            case 1:
+                return self.real.x = value;
+            case 2:
+                return self.real.y = value;
+            case 3:
+                return self.real.z = value;
+            case 4:
+                return self.dual.s = value;
+            case 5:
+                return self.dual.x = value;
+            case 6:
+                return self.dual.y = value;
+            case 7:
+                return self.dual.z = value;
+            default:
+                throw "Invalid element";
+        }
+    }
+    
+    /**
+     * Rotate the given vector, assuming the current DualQuaternion is normalized (if not, normalize first).
+     * 
+     * @param u     The vector to rotate.
+     * @return      The rotated vector.
+     */
+    public inline function rotate(u:Vector3):Vector3
+    {
+        normalize();
+        return this.real.rotate(u);
+    }
+    
+    
+    /**
+     * Find the arccosine of the angle between two rotation quaternions in the DualQuaternion
+     * 
+     * @param b     The other quaternion.
+     * @return      The arccosine angle between this vector and the other in radians.
+     */
+    public inline function angleWith(b:Quaternion):Float
```

Comment:
This seems unnecessary. There's an angleWith on Quaternion, so the caller can just call dualQuaternion.real.angleWith(quaternion).